src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php

```diff
@@ -24,9 +24,15 @@
  */
 interface DenormalizerInterface
 {
+    const COLLECT_INVARIANT_VIOLATIONS = 'collect_invariant_violations';
+
     /**
      * Denormalizes data back into an object of the given class.
      *
+     * When context option `collect_invariant_violations` is enabled, the
```

Comment:
Could we deprecate not enabling this option? So in 6.0 we will have only one code path again.

Also, shouldn't we add a similar API to `NormalizerInterface` for consistency?

@symfony/mergers I'm interested in your opinion about this PR because it's an important change in the API.

Reply:
> Could we deprecate not enabling this option? So in 6.0 we will have only one code path again.

I thought it would be better to have this as an experimental feature so as users don't get forced into a big API change. But thinking about it again, keeping it as an opt-in feature probably means third-party normalizers won't make a move to be compatible with the new API (because maintainers don't know about it or they can't/don't want to update). Deprecating the current API in favor of this one will also allow enforcing the API at type level by making `DenormalizerInterface::denormalize()` return a `DenormalizationResult`.

What I'm affraid of is that the new API might introduce some overhead to the denormalization process, is that something we want to be the new default?

Also: if the new API is the only one in 6.0, this new option shall be deprecated in 6.0 and removed in 7.0, right?

> Also, shouldn't we add a similar API to NormalizerInterface for consistency?

The use case of the new API for denormalization is that input data can be untrusted and one might want to show detailed error messages rather than a technical exception to the user that provided that input. Is there a similar scenario for normalization? I'm totally ok to apply the same API changes to normalization, I just want to be sure it's actually worth it.

Reply:
It's ok to deprecate and 5.x and remove in 6.

I'm not sure about the use cases for `NormalizerInterface`, but I'm sure we can find some. For instance in API Platform we often have to store metadata along with the normalization result. An example: we store the IRIs of visited documents to generate cache tags. We use currently use mighty tricks involving the serialization context, having a result object would allow to make cleaner things. We'll probably need an extension point for this result object by the way (it can start as a simple a context map).

That being said, the key point here is consistency IMHO. Normalization and de normalization must work in a similar way. Consistent APIs are easier to learn and remember.